plugins/repository-s3/src/main/java/org/elasticsearch/repositories/s3/S3BlobStore.java

```diff
@@ -51,15 +52,18 @@
 
     private final StorageClass storageClass;
 
-    S3BlobStore(S3Service service, String clientName, String bucket, boolean serverSideEncryption,
-                ByteSizeValue bufferSize, String cannedACL, String storageClass) {
+    private final Tuple<RepositoryMetaData, Settings> settingsKey;
```

Comment:
this is perhaps a bit too coarse-grained as settings key? Should it just be repository metadata that's relevant for establishing connections?

Also, why are the settings (i.e. 2nd part of tuple) part of this key? These are static / fixed at node startup, so there's no need to cache them?

Reply:
@ywelsch 

> Also, why are the settings (i.e. 2nd part of tuple) part of this key? These are static / fixed at node startup, so there's no need to cache them?

I think I was a little paranoid here with reloads from the secret store but on second thought, that's wrong it's all handled by `refreshAndClearCache` just fine :). I'll simplify the key.

> this is perhaps a bit too coarse-grained as settings key? Should it just be repository metadata that's relevant for establishing connections?

Maybe in theory, but does it matter in practice? Creating a new repo client isn't absurdly expensive and updating the repo settings is a rare event? I agree we could create another key here but it seems like it's just a bunch of extra code complexity and then if we ever add a new relevant (for establishing the connection) setting we have to maintain this as well => maybe the rare case of needlessly creating a new client isn't a problem?

Reply:
> Maybe in theory, but does it matter in practice?

I was wondering if with this approach we never share clients between repo definitions (as these repos will probably be different in some way). Not sure how many repo definitions we typically have in clusters though and whether it matters.

Reply:
Wait :) I can't take credit for it now, but I think I actually handled this just fine already :D 
The actual client instances are cached by their `S3ClientSettings` which only contain stuff needed for instantiating the clients.
So `org.elasticsearch.repositories.s3.S3Service#clientsCache` should deduplicate clients just fine. All we get are some redundant settings instances in `org.elasticsearch.repositories.s3.S3Service#derivedClientSettings` where two different repo-metadata keys could point to equal settings instances.